### DISCUSSION TITLE
[Pro] HTML in Select2 fields

### DISCUSSION THREAD

**Discussion by giovdi** [Laravel-Backpack/community-forum#1109] [Ideas]
Hi everyone!
Today, I was wondering if it's possible to write HTML into a Select2 dropdown with Backpack. Well, the short answer (I think) is no... but...

I tried looking at the code that generates the Select2 Javascript for my specific case, and it is probably pretty easy to do.

In this example, I added the following code blocks to `backpack/pro/resources/views/fields/relationship/select.blade.php`.

```diff
    // make sure the $field['value'] takes the proper value
    $current_value = old_empty_or_null($field['name'], []) ??  $field['value'] ?? $field['default'] ?? [];

+   if(isset($field['html']) && !is_array($field['html'])) {
+       $field['html'] = [true];
+   }
+   $activeHtml = !empty($field['html']) ? true : false;
+
+   if ($activeHtml) {
+       $field['html']['base_element'] = $field['html']['base_element'] ?? 'span';
+   }

    if (!empty($current_value) || is_int($current_value)) {
        switch (gettype($current_value)) {
```

```diff
        @if($field['multiple'])
        multiple
        @endif

+       data-html="{{ var_export($field['allows_null']) }}"
+       @if($activeHtml)
+         data-html-element="{{ $field['html']['base_element'] }}"
+       @endif
        >
```

```diff
    function bpFieldInitRelationshipSelectElement(element) {
        const $placeholder = element.attr('data-placeholder');
        const $multiple = element.attr('data-field-multiple')  == 'false' ? false : true;
+       const $html = element.attr('data-html')  == 'true' ? true : false;
+       const $htmlElement = element.attr('data-html-element');
        const $allows_null = (element.attr('data-column-nullable') == 'true') ? true : false;
```

```diff
        var $select2Settings = {
                theme: 'bootstrap',
                multiple: $multiple,
                placeholder: $placeholder,
                allowClear: $allowClear,
                dropdownParent: $isFieldInline ? $('#inline-create-dialog .modal-content') : $(document.body)
            };
+           if ($html) {
+               $select2Settings.templateSelection = function(elem) {
+                   if (!elem.id) {
+                       return elem.text;
+                   }
+
+                   return $('<'+$htmlElement+'>'+elem.text+'</'+$htmlElement+'>');
+               }
+               $select2Settings.templateResult = function(elem) {
+                   if (!elem.id) {
+                       return elem.text;
+                   }
+
+                   return $('<'+$htmlElement+'>'+elem.text+'</'+$htmlElement+'>');
+               }
+           }
        if (!$(element).hasClass("select2-hidden-accessible"))
```

At this point, I can add the following option to the CRUD:

```diff
CRUD::field([
	'name' => 'variants',
	'type' => "relationship",
	'inline_create' => true,
	'init_rows' => 1,
	'min_rows' => 1,
	'subfields' => [
		[
			'name' => 'price',
			'label' => 'Price',
			'type' => 'number',
			'prefix' => "&euro;",
			'attributes' => ["step" => "any"],
			'wrapper' => [
				'class' => 'form-group col-md-3'
			],
		],
		[
			'name' => 'weight_initial',
			'label' => 'Weight',
			'type' => 'number',
			'suffix' => "grams",
			'wrapper' => [
				'class' => 'form-group col-md-3'
			],
		],
	],
	'pivotSelect' => [
+		'html' => true,
		'wrapper' => [
			'class' => 'form-group col-md-6'
		],
		'options' => (function ($query) {
			return $query->orderBy('variant_name', 'ASC')->get();
		})
	],
]);
```

...and with the following accessor:
```php
public function identifiableAttribute()
{
	return 'variant_name_full';
}

protected function variantNameFull(): Attribute
{
	return Attribute::make(
		get: function () {
			$ret = '';
			if (!empty($this->product->filament_material)) {
				$ret .= '<span class="badge badge-primary">'.$this->product->filament_material->filament_material.'</span> ';
			}
			if (!empty($this->attributes['color'])) {
				$ret .= '<span class="badge" style="border:1px solid black; background-color:'.$this->attributes['color'].'">&nbsp;</span>';
			}
			$ret .= " ".$this->product->manufacturer->manufacturer_name . ' - ' . $this->product->product_name . ' - ' . $this->attributes['variant_name'];
			if (!empty($this->code)) {
				$ret .= ' ('.$this->code.')';
			}
			return $ret;

		},
	);
}
```

...here is the result!
<img width="914" alt="Screenshot 2024-08-07 222805" src="https://github.com/user-attachments/assets/d9f70324-23c7-46d4-8a63-68f00b844444">

Not so bad, isn't it?

As you can see from the code blocks, a `<span>` tag surrounds the HTML option, but you can customize it with the `base_element` property, as follows: `'html' => ['base_element' => 'div']`.
You can't avoid this because if you mix plain text options and HTML ones, the `$(elem.text);` fails with a JS error, so at least one HTML tag is required.

Please let me know your suggestions... or if I just lost 2 hours of my life for some option that I don't know 😄 
Unfortunately, I don't have enough time to create a full PR in a reasonable time, so please, if it's a good idea, take care of it :)

Thanks a lot!

**Comment by pxpm** [comment]
Hey @giovdi thank a lot for the suggestion. 

I really think this opens up the possibility to some really cool features, like showing avatars on user select lists etc. I am willing to give this a shot, there are some un-handled scenarios in your example that I am worried it could not work as I am expecting, for example the translatable selects. 

Can't promise any deadline, but I am moving this to the CRUD repository as a feature request because if it goes smooth on the scenarios we haven't tested yet, it will be an "easy" win for everyone. 👍 

Once again, thanks for the suggestion. 🙏 

Cheers

**Comment by giovdi** [comment]
Hi @pxpm!
Thank you for your feedback! :)

Take your time to look over all the possibilities and scenarios (and potential issues). I know that this can take months; there's no rush.
For example, I tried to do this to a Select2 outside a relationship, and the changes to make (based on the previous message) are similar but slightly different.

Cheers!
